src/lib.rs

```diff
@@ -3981,11 +3981,21 @@ pub unsafe trait IntoBytes {
     where
         Self: Immutable,
     {
-        if bytes.len() != mem::size_of_val(self) {
-            return Err(SizeError::new(self));
+        if bytes.len() == mem::size_of_val(self) {
+            // SAFETY: Within this branch of the conditional, we have ensured
+            // that `bytes.len()` (the size of the destination) is equal to
+            // `mem::size_of_val(self)` the size of the source value. Neither
+            // the size of the source nor the size of the destination change
+            // between the above size check and the invocation of
+            // `copy_unchecked`. The size of `self` (the source) cannot change,
+            // because `Self` is bounded as `Immutable`, and the size of `bytes`
+            // cannot change, because we hold an exclusive reference to it and
+            // do not change its size.
+            unsafe { util::copy_unchecked(self.as_bytes(), bytes) }
```

Comment:
```suggestion
        let self_bytes = self.as_bytes();
        if bytes.len() == self_bytes.len() {
            // SAFETY: Within this branch of the conditional, we have ensured
            // that `bytes.len()` (the size of the destination) is equal to
            // `self_bytes.len()` (the size of the source). Neither
            // the size of the source nor the size of the destination change
            // between the above size check and the invocation of
            // `copy_unchecked`.
            unsafe { util::copy_unchecked(self_bytes, bytes) }
```

(And reflow the safety commnet)

This makes the safety argument simpler since we don't need to reason about the relationship between `size_of_val(self)` and `self.as_bytes()`. It is probably ever so slightly tougher on the optimizer, but I doubt it's enough to make a difference.

Reply:
[Done](https://github.com/google/zerocopy/compare/6e1ee44e19fac8be1e2122528c3d88ebbbd00cd0..78d9f495abc6235b27f4d76b5f3c7e299990ef3e#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R3984)